plugins/block-sharable-procedures/src/observable_parameter_model.ts

```diff
@@ -0,0 +1,129 @@
+      varId?: string) {
+    this.id = id ?? Blockly.utils.idGenerator.genUid();
+    this.variable = this.workspace.getVariable(name) ??
+        workspace.createVariable(name, '', varId);
+  }
+
+  /**
+   * Sets the name of this parameter to the given name.
+   * @param name The string to set the name to.
+   * @return This parameter model.
+   */
+  setName(name: string): this {
+    if (name === this.variable.name) return this;
+    const oldName = this.variable.name;
+    this.variable =
+        this.workspace.getVariable(name) ?? this.workspace.createVariable(name);
+    triggerProceduresUpdate(this.workspace);
+    if (this.shouldFireEvents) {
+      // Blockly.Events.fire(
+      //     new (Blockly.Events.get(
+      //         Blockly.Events.PROCEDURE_PARAMETER_RENAME))(
+      //         this.workspace, this.procedureModel, this, oldName));
+    }
+    return this;
+  }
+
+  /**
+   * Unimplemented. The built-in ParameterModel does not support typing.
+   * If you want your procedure blocks to have typed parameters, you need to
+   * implement your own ParameterModel.
+   * @param types The types to set this parameter to.
+   * @throws Throws for the ObservableParameterModel specifically because this
+   *     method is unimplemented.
+   */
+  setTypes(types: string[]): this {
+    throw new Error(
+        'The built-in ParameterModel does not support typing. You need to ' +
+        'implement your own custom ParameterModel.');
+  }
+
+  /**
+   * Returns the name of this parameter.
+   * @return the name of this parameter.
+   */
+  getName(): string {
+    return this.variable.name;
+  }
+
+  /**
+   * Returns the types of this parameter.
+   * @return the types of this parameter.
+   */
+  getTypes(): string[] {
+    return [];
+  }
+
+  /**
+   * Returns the unique language-neutral ID for the parameter. This represents
+   * the identity of the variable model which does not change over time.
+   * @return The unique language-neutral ID for the parameter.
+   */
+  getId(): string {
+    return this.id;
+  }
+
+  /**
+   * Returns the variable model associated with the parameter model.
+   * @return the variable model associated with the parameter model.
+   */
+  getVariableModel(): Blockly.VariableModel {
+    return this.variable;
+  }
+
+  /**
+   * Tells the parameter model it should fire events.
+   * @internal
+   */
+  startPublishing() {
+    this.shouldFireEvents = true;
+  }
+
+  /**
+   * Tells the parameter model it should not fire events.
+   * @internal
+   */
+  stopPublishing() {
+    this.shouldFireEvents = false;
+  }
+
+  /**
+   * Sets the procedure model this parameter is a part of.
+   * @param model The procedure model this parameter is a part of.
+   * @return This parameter model.
+   */
+  setProcedureModel(model: Blockly.procedures.IProcedureModel): this {
+    // TODO: Not sure if we want to do this, or accept it via the constructor.
```

Comment:
Making it be non-null seems better / developer can't forget this step and get weird bugs.

But I don't have a good sense of how the fluent API would be used, do you have an example?

Reply:
So right now the API works like this (which is fluent):

```
const procedure = new ObservableProcedureModel('my procedure name')
    .setReturnTypes(['Number'])
    .insertParameter(new ObservableParameterModel('my param name'), 0);
```

While if we required the `ObservableProcedureModel` to be passed directly, it would have to look like this (which is not fluent):

```
const procedure = new ObservableProcedureModel('my procedure name')
    .setReturnTypes(['Number']);
 procedure.insertParameter(new ObservableParameterModel('my param name', procedure), 0);
```

I think it is better to preserve the fluent API in this case. After building it out, I don't think there's really a risk to making the `procedureModel` nullable, because the `ObservableProcedureModel` handles calling `setProcedureModel` inside `insertParameter`. External developers don't have to remember to call it.